### ISSUE TITLE
Plugin causes idea EDT Freeze

### ISSUE THREAD

**Issue by Gr33nbl00d** [asciidoctor/asciidoctor-intellij-plugin#584]
When editing normal Java files the IDE freezes sometimes.
By analyzing the threaddump generated by idea i found the problem is related to this plugin.

 on com.intellij.openapi.application.impl.ReadMostlyRWLock@2d158470
	at java.base@11.0.8/jdk.internal.misc.Unsafe.park(Native Method)
	at java.base@11.0.8/java.util.concurrent.locks.LockSupport.parkNanos(LockSupport.java:234)
	at com.intellij.openapi.application.impl.ReadMostlyRWLock.writeLock(ReadMostlyRWLock.java:251)
	at com.intellij.openapi.application.impl.ApplicationImpl.acquireWriteLock(ApplicationImpl.java:894)
	at com.intellij.openapi.application.impl.ApplicationImpl.startWrite(ApplicationImpl.java:1175)
	at com.intellij.openapi.application.impl.ApplicationImpl.runWriteAction(ApplicationImpl.java:978)
	at com.intellij.codeInsight.editorActions.PasteHandler.doPaste(PasteHandler.java:180)
	at com.intellij.codeInsight.editorActions.PasteHandler.lambda$execute$2(PasteHandler.java:102)
	at com.intellij.codeInsight.editorActions.PasteHandler$$Lambda$4493/0x0000000102bdb840.run(Unknown Source)
	at com.intellij.openapi.project.DumbService.runWithAlternativeResolveEnabled(DumbService.java:368)
	at com.intellij.codeInsight.editorActions.PasteHandler.execute(PasteHandler.java:93)
	at org.asciidoc.intellij.actions.editorAction.AsciiDocPasteImageHandler.execute(AsciiDocPasteImageHandler.java:70)
	at org.asciidoc.intellij.actions.editorAction.AsciiDocPasteImageHandler.doExecute(AsciiDocPasteImageHandler.java:43)
	at com.intellij.openapi.editor.actionSystem.DynamicEditorActionHandler.doExecute(DynamicEditorActionHandler.java:52)
	at com.intellij.openapi.editor.actionSystem.EditorActionHandler.lambda$execute$4(EditorActionHandler.java:198)
	at com.intellij.openapi.editor.actionSystem.EditorActionHandler$$Lambda$3830/0x0000000102920840.perform(Unknown Source)
	at com.intellij.openapi.editor.actionSystem.EditorActionHandler.doIfEnabled(EditorActionHandler.java:88)
	at com.intellij.openapi.editor.actionSystem.EditorActionHandler.execute(EditorActionHandler.java:197)
	at com.intellij.openapi.editor.actionSystem.EditorAction.lambda$actionPerformed$0(EditorAction.java:89)
	at com.intellij.openapi.editor.actionSystem.EditorAction$$Lambda$3829/0x0000000102920440.run(Unknown Source)
	at com.intellij.openapi.command.impl.CoreCommandProcessor.executeCommand(CoreCommandProcessor.java:220)
	at com.intellij.openapi.command.impl.CoreCommandProcessor.executeCommand(CoreCommandProcessor.java:177)
	at com.intellij.openapi.editor.actionSystem.EditorAction.actionPerformed(EditorAction.java:98)
	at com.intellij.openapi.editor.actionSystem.EditorAction.actionPerformed(EditorAction.java:73)
	at com.intellij.openapi.actionSystem.ex.ActionUtil.performActionDumbAware(ActionUtil.java:282)
	at com.intellij.openapi.keymap.impl.IdeKeyEventDispatcher$1.performAction(IdeKeyEventDispatcher.java:610)
	at com.intellij.openapi.keymap.impl.IdeKeyEventDispatcher.lambda$processAction$3(IdeKeyEventDispatcher.java:670)
	at com.intellij.openapi.keymap.impl.IdeKeyEventDispatcher$$Lambda$3828/0x0000000102920040.run(Unknown Source)
	at com.intellij.openapi.application.TransactionGuardImpl.performUserActivity(TransactionGuardImpl.java:94)
	at com.intellij.openapi.keymap.impl.IdeKeyEventDispatcher.processAction(IdeKeyEventDispatcher.java:669)
	at com.intellij.openapi.keymap.impl.IdeKeyEventDispatcher.processAction(IdeKeyEventDispatcher.java:620)
	at com.intellij.openapi.keymap.impl.IdeKeyEventDispatcher.processActionOrWaitSecondStroke(IdeKeyEventDispatcher.java:517)
	at com.intellij.openapi.keymap.impl.IdeKeyEventDispatcher.inInitState(IdeKeyEventDispatcher.java:471)
	at com.intellij.openapi.keymap.impl.IdeKeyEventDispatcher.dispatchKeyEvent(IdeKeyEventDispatcher.java:220)
	at com.intellij.ide.IdeEventQueue.dispatchKeyEvent(IdeEventQueue.java:890)
	at com.intellij.ide.IdeEventQueue._dispatchEvent(IdeEventQueue.java:835)
	at com.intellij.ide.IdeEventQueue.lambda$dispatchEvent$8(IdeEventQueue.java:452)
	at com.intellij.ide.IdeEventQueue$$Lambda$517/0x0000000100640c40.compute(Unknown Source)
	at com.intellij.openapi.progress.impl.CoreProgressManager.computePrioritized(CoreProgressManager.java:744)
	at com.intellij.ide.IdeEventQueue.lambda$dispatchEvent$9(IdeEventQueue.java:451)
	at com.intellij.ide.IdeEventQueue$$Lambda$516/0x000000010061b440.run(Unknown Source)
	at com.intellij.openapi.application.impl.ApplicationImpl.runIntendedWriteActionOnCurrentThread(ApplicationImpl.java:802)
	at com.intellij.ide.IdeEventQueue.dispatchEvent(IdeEventQueue.java:505)
	at java.desktop@11.0.8/java.awt.EventDispatchThread.pumpOneEventForFilters(EventDispatchThread.java:203)
	at java.desktop@11.0.8/java.awt.EventDispatchThread.pumpEventsForFilter(EventDispatchThread.java:124)
	at java.desktop@11.0.8/java.awt.EventDispatchThread.pumpEventsForHierarchy(EventDispatchThread.java:113)
	at java.desktop@11.0.8/java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:109)
	at java.desktop@11.0.8/java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:101)
	at java.desktop@11.0.8/java.awt.EventDispatchThread.run(EventDispatchThread.java:90)

**Comment by Gr33nbl00d** [comment]
I think the problem is inside AsciiDocPasteImageHandler
Not analyzed this deeply but you should check first if there is any processing needed at all.
If the edited file is an ascii doc file at all. And if the pasted data is really image data. These checks should be done before any kind of processing time is spend.

**Comment by ahus1** [comment]
Hi @Gr33nbl00d, I'm sorry to hear your IDE blocks. 

The AsciiDocPasteImageHandler intercepts paste, but then quickly decides to pass the call to the original paste handler when it doesn't have anything to do. It still appears in the stack trace, but the lock occurs in the real paste handler when it tries to acquire the write lock. 

Can you provide the full thread dump (maybe here, maybe in a separate email directly to me - you find my email address in my profile). Looking at the thread dump I hope to see who currently holds the write lock while this thread waits for it.

Thanks!

**Comment by ahus1** [comment]
No thread dump yet, so I close this issue. Will re-open it once more information becomes available.

**Comment by ahus1** [comment]
The next upcoming release will no longer implement AsciiDocPasteImageHandler as it is no longer needed in the latest IntelliJ version. This should avoid this plugin to appear in EDT freezes.

**Comment by ahus1** [comment]
Pre-release 0.32.5 is now available.

The pre-release of the plugin is available from [GitHub releases](https://github.com/asciidoctor/asciidoctor-intellij-plugin/releases) and the [IntelliJ AsciiDoc EAP repository](https://intellij-asciidoc-plugin.ahus1.de/docs/contributors-guide/user/add-eap-repository-to-ide.html).
